File: day-1/rust/src/solution/not_quite_lisp.rs
const UP_FLOOR: char = '(';
const DOWN_FLOOR: char = ')';

pub fn part_1(directions: &str) -> Result<isize, String> {
  let mut floor = 0;

  for direction in directions.trim().chars() {
    match direction {
      UP_FLOOR => floor += 1,
      DOWN_FLOOR => floor -= 1,
      _ => return Err(format!("Invalid direction: {}", direction)),
    }
  }

  Ok(floor)
}

pub fn part_2(directions: &str) -> Result<usize, String> {
  let mut floor = 0;

  for (idx, direction) in directions.trim().chars().enumerate() {
    match direction {
      UP_FLOOR => floor += 1,
      DOWN_FLOOR => {
        floor -= 1;
        if floor < 0 {
          return Ok(idx + 1)
        }
      },
      _ => return Err(format!("Invalid direction: {}", direction)),
    }
  }

  Ok(0)
}

File: day-1/go/solution/not_quite_lisp.go
package not_quite_lisp

type NotQuiteLisp interface {
	Part1(data string) int64
	Part2(data string) int64
}

type notQuiteLisp struct{}

func New() NotQuiteLisp {
	return &notQuiteLisp{}
}

const UP_FLOOR rune = '('
const DOWN_FLOOR rune = ')'

func (n *notQuiteLisp) Part1(data string) int64 {
	counter := int64(0)
	for _, char := range data {
		switch char {
		case DOWN_FLOOR:
			counter -= 1
		case UP_FLOOR:
			counter += 1
		}
	}
	return counter
}

func (n *notQuiteLisp) Part2(data string) int64 {
	counter := int64(0)

	for idx, char := range data {
		switch char {
		case UP_FLOOR:
			counter += 1
		case DOWN_FLOOR:
			counter -= 1
		}
		if counter < 0 {
			return int64(idx + 1)
		}
	}

	return -1
}
